hyperspy/misc/io/fei_stream_readers.py

```diff
@@ -0,0 +1,320 @@
+                break
+        # if different of ‘65535’, add a count to the corresponding channel
+        if value != 65535:  # Same spectrum
+            if data:
+                if value == count_channel:  # Same channel, add a count
+                    data += 1
+                else:  # a new channel, same spectrum—requires new coord
+                    # Store previous channel
+                    coords.append((
+                        int(navigation_index // xsize),
+                        int(navigation_index % xsize),
+                        int(count_channel // rebin_energy))
+                    )
+                    data_list.append(data)
+                    # Add a count to new channel
+                    data = 1
+                    # Update count channel as this is a new channel
+                    count_channel = value
+
+            else:  # First non-zero channel of spectrum
+                data = 1
+                # Update count channel as this is a new channel
+                count_channel = value
+
+        else:  # Advances one pixel
+            if data:  # Only store coordinates if the spectrum was not empty
+                coords.append((
+                    int(navigation_index // xsize),
+                    int(navigation_index % xsize),
+                    int(count_channel // rebin_energy))
+                )
+                data_list.append(data)
+            navigation_index += 1
+            data = 0
+
+    final_shape = (ysize, xsize, channels // rebin_energy)
+    coords = np.array(coords).T
+    data = np.array(data_list)
+    return coords, data, final_shape
+
+
+@jit_ifnumba
+def _stream_to_sparse_COO_array(
+        stream_data, last_frame, shape, channels, rebin_energy=1, first_frame=0):
+    navigation_index = 0
+    frame_number = 0
+    ysize, xsize = shape
+    frame_size = xsize * ysize
+    data_list = []
+    coords = []
+    data = 0
+    count_channel = None
+    for value in stream_data:
+        if frame_number < first_frame:
+            if value != 65535:  # Same spectrum
+                continue
+            else:
+                navigation_index += 1
+                if navigation_index == frame_size:
+                    frame_number += 1
+                    navigation_index = 0
+                continue
+        # when we reach the end of the frame, reset the navigation index to 0
+        if navigation_index == frame_size:
+            navigation_index = 0
+            frame_number += 1
+            if frame_number == last_frame:
+                break
+        # if different of ‘65535’, add a count to the corresponding channel
+        if value != 65535:  # Same spectrum
+            if data:
+                if value == count_channel:  # Same channel, add a count
+                    data += 1
+                else:  # a new channel, same spectrum—requires new coord
+                    # Store previous channel
+                    coords.append((
```

Comment:
Does it matter when using numba? It may actually decrease the performance in this case. Also, I am not sure of where you are proposing to store the coordinates in between the expansion steps.

Reply:
yeah, maybe I guess this is already good enough for now!